indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/NumberEncoding.kt

```diff
@@ -241,222 +163,374 @@ fun Int.toUnsignedByteArray() = toLong().toUnsignedByteArray()
+ * Decodes an ASN.1 unsigned varint to an [UInt], copying all bytes from the source into a [ByteString].
  *
- * @return the decoded ULong and the underlying varint-encoded bytes as `ByteArray`
- * @throws IllegalArgumentException if the number is larger than [ULong.MAX_VALUE]
+ * @return the decoded [UInt] and the underlying varint-encoded bytes as ByteString
+ * @throws IllegalArgumentException if the number is larger than [UInt.MAX_VALUE]
  */
-inline fun ByteArray.decodeAsn1VarULong(): Pair<ULong, ByteArray> = iterator().decodeAsn1VarULong()
+@Throws(IllegalArgumentException::class)
+fun Source.decodeAsn1VarUInt(): Pair<UInt, ByteArray> =
+    decodeAsn1VarInt(UInt.SIZE_BITS).let { (n, b) -> n.toUInt() to b }
 
 /**
- * Decodes an ULong from bytes using varint encoding as used within ASN.1: groups of seven bits are encoded into a byte,
- * while the highest bit indicates if more bytes are to come. Trailing bytes are ignored.
+ * Decodes an ASN.1 unsigned varint to an ULong allocating at most [bits] many bits .
+ * This function is useful as an intermediate processing step, since it also returns a [Buffer]
+ * holding all bytes consumed from the source.
+ * This operation essentially moves bytes around without copying.
  *
- * @return the decoded ULong and the underlying varint-encoded bytes as `ByteArray`
- * @throws IllegalArgumentException if the number is larger than [ULong.MAX_VALUE]
+ * @return the decoded ASN.1 varint as an [ULong] and the underlying varint-encoded bytes as [Buffer]
+ * @throws IllegalArgumentException if the resulting number requires more than [bits] many bits to be represented
  */
-fun Iterator<Byte>.decodeAsn1VarULong(): Pair<ULong, ByteArray> {
+@Throws(IllegalArgumentException::class)
+private fun Source.decodeAsn1VarInt(bits: Int): Pair<ULong, ByteArray> {
     var offset = 0
     var result = 0uL
-    val accumulator = mutableListOf<Byte>()
-    while (hasNext()) {
-        val current = next().toUByte()
-        accumulator += current.toByte()
-        if (current >= 0x80.toUByte()) {
-            result = (current and 0x7F.toUByte()).toULong() or (result shl 7)
+    val accumulator = Buffer()
+    while (!exhausted()) {
+        val current = readUByte()
+        accumulator.writeUByte(current)
+        if (current >= UVARINT_SINGLEBYTE_MAXVALUE_UBYTE) {
+            result = (current and UVARINT_MASK_UBYTE).toULong() or (result shl 7)
         } else {
-            result = (current and 0x7F.toUByte()).toULong() or (result shl 7)
+            result = (current and UVARINT_MASK_UBYTE).toULong() or (result shl 7)
             break
         }
-        if (++offset > ceil(ULong.SIZE_BYTES.toFloat() * 8f / 7f)) throw IllegalArgumentException("Tag number too Large do decode into ULong!")
+        if (++offset > ceil((bits * 8).toFloat() * 8f / 7f)) throw IllegalArgumentException("Tag number too Large do decode into $bits bits!")
```

Comment:
this is a generic asn.1 varint decoder, the exception message should not mention tags

also `do decode` -> `to decode`